<span style="font-size:30px;">**1** Load Dataset</span>

In [10]:
import numpy as np
import pandas as pd
from numpy import loadtxt
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [11]:
dataset = pd.read_csv("/Users/albertegi/Dev/Diabetes_Prediction_Using_ANN/sample_data/diabetes.csv")
print(dataset.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  


In [12]:
X = dataset.iloc[:, 0:8]  # Select all rows and columns 0 to 7
y = dataset.iloc[:, 8]    # Select column 8 (target variable)

In [13]:
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

<span style="font-size:30px;">**2** Define the Model</span>
### This Neural Network is a fully connected network with three layers
1. The model expects rows of data with 8 variables (the input_dim = 8 argument)
2. The first hidden layer has 12 nodes and uses the relu activation function
3. The second hidden layer 9 nodes and uses the relu activation function
4. The output layer has one node and uses the sigmoid activation function.

 **Note** That the shape of the input to the model is defined as an argument on the first hidden layer
This means that the line of code that adds the first Dense layer does two things. Defining the input and the first hidden layer

In [14]:
model = Sequential()
model.add(Dense(units=12, input_dim=8, activation='relu')) # input dimension and first layer
model.add(Dense(units=9, activation='relu')) # second layer
model.add(Dense(units=1,activation='sigmoid')) # third layer also called the output layer

# model = Sequential([
#     Input(shape=(8,)),  # Define input shape explicitly
#     Dense(units=12, activation='relu'),  # First hidden layer
#     Dense(units=9, activation='relu'),   # Second hidden layer
#     Dense(units=1, activation='sigmoid') # Output layer
# ])

/Users/albertegi/Dev/Diabetes_Prediction_Using_ANN/venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


<span style="font-size:30px;">**3** Compile the Model</span> 

Now that the model is defined, we can compile it (i.e. we need to put all our recipe
that we need together in order to get ready for training).

We must specify the loss function to use to evaluate a set of weights, the optimizer is is used to search
through different weights for the network and any optional metrics we would like to collect and report during training.
in this case, we will use cross entropy as the loss argument. this loss is for a binary classification problems and is defined in keras as **binary_crossentropy.**

We will define the optimizer as the efficient stochastic gradient descent algorithm **"adam".** This is a popular version of of gradient descent because it automatically tunes itself and gives good result in a wide range of problems.

Finally, because it is a classification problem, we will collect and report the classification accuracy via the mtrics argument.

In [15]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

<span style="font-size:30px;">**4** Fit the Model</span> 

We have defined our model and compiled it ready for efficient training.
Now it is time to execute the model on some data. We can train or fit our model on our loaded data by calling the **fit** function on the model.

The training occurs over epochs and each epoch is split into batches.
1. **Epoch:** one pass through all of the rows in the training dataset
2. **Batch:** One or more samples considered by the model within an epoch before weights are updated.
For this problem, we will run for a small number of epochs (150) and use a relatively small batch size of 10
                                               

In [16]:
model.fit(X, y, epochs=150, batch_size=10)

Epoch 1/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 394us/step - accuracy: 0.6464 - loss: 46.6906  
Epoch 2/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 318us/step - accuracy: 0.5957 - loss: 7.2884
Epoch 3/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 308us/step - accuracy: 0.5174 - loss: 3.0207
Epoch 4/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 307us/step - accuracy: 0.5557 - loss: 1.7789
Epoch 5/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 311us/step - accuracy: 0.5528 - loss: 1.2525
Epoch 6/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 304us/step - accuracy: 0.5806 - loss: 0.9887
Epoch 7/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 311us/step - accuracy: 0.6177 - loss: 0.7784
Epoch 8/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 324us/step - accuracy: 0.6021 - loss: 0.7868
Epoch 9/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 322us/step - accuracy: 0.6517 - loss: 0.6818
Epoch 10/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 316us/step - accuracy: 0.6573 - loss: 0.6397
Epoch 11/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 309us/step - accuracy: 0.6755 - loss: 0.7466
Epoch 12/150
77/77 ━━━━━━━━━━━━━━━━━━━

We can see that we have 768 records, so out of that, we choose 10 records at a time (referred to as batch size) for 1 iteration (referred to as one epoch)

<span style="font-size:30px;">**5** Evaluate Keras Model</span> 

We have trained our neural network on the entire dataset and we can evaluate the performance of the network on the same dataset.
You can evaluate your model on your training dataset using the evaluate() function on your model and pass it the same input and output used to train the model.

This will generate a prediction for each input and output pair and collect scores, including the average loss and any metrics you have configured, such as accuracy.

The evaluate function will return a list with two values. the first will be the loss of the model on the dataset and the second will be the accuracy of the model on the dataset

In [20]:
loss, accuracy = model.evaluate(X, y)
print("Accuracy: {}\n Loss: {} ".format(accuracy*100, loss))

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7155 - loss: 0.5385
Accuracy: 74.609375
 Loss: 0.5182375311851501 


<span style="font-size:30px;">**6** Make Predictions</span> 

Making predictions is as easy as calling the predict() function on the model. We are using a sigmoid activation function on the output layer, so the predictions will be probability in the range between 0 and 1. we can easily convert them into a crisp binary prediction for this classification task by round them.

In [22]:
predictions = model.predict(X)
rounded = [round(X[0]) for X in predictions]
rounded

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


[1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,


In [25]:
# Get predictions (probabilities) from the model
predictions_prob = model.predict(X)

# Convert probabilities to binary predictions (0 or 1) using a threshold of 0.5
predictions = (predictions_prob > 0.5).astype(int)

# Summarize the first 20 cases
for i in range(20):
    print(f'Input: {X.iloc[i].tolist()} => Predicted: {predictions[i]} (Expected: {y.iloc[i]})')

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 455us/step
Input: [6.0, 148.0, 72.0, 35.0, 0.0, 33.6, 0.627, 50.0] => Predicted: [1] (Expected: 1)
Input: [1.0, 85.0, 66.0, 29.0, 0.0, 26.6, 0.351, 31.0] => Predicted: [0] (Expected: 0)
Input: [8.0, 183.0, 64.0, 0.0, 0.0, 23.3, 0.672, 32.0] => Predicted: [1] (Expected: 1)
Input: [1.0, 89.0, 66.0, 23.0, 94.0, 28.1, 0.167, 21.0] => Predicted: [0] (Expected: 0)
Input: [0.0, 137.0, 40.0, 35.0, 168.0, 43.1, 2.288, 33.0] => Predicted: [1] (Expected: 1)
Input: [5.0, 116.0, 74.0, 0.0, 0.0, 25.6, 0.201, 30.0] => Predicted: [0] (Expected: 0)
Input: [3.0, 78.0, 50.0, 32.0, 88.0, 31.0, 0.248, 26.0] => Predicted: [0] (Expected: 1)
Input: [10.0, 115.0, 0.0, 0.0, 0.0, 35.3, 0.134, 29.0] => Predicted: [1] (Expected: 0)
Input: [2.0, 197.0, 70.0, 45.0, 543.0, 30.5, 0.158, 53.0] => Predicted: [1] (Expected: 1)
Input: [8.0, 125.0, 96.0, 0.0, 0.0, 0.0, 0.232, 54.0] => Predicted: [0] (Expected: 1)
Input: [4.0, 110.0, 92.0, 0.0, 0.0, 37.6, 0.191, 30.0] => Predicted: [0] (Expecte